### PR TITLE
fix: retain the original line break style of the file and fix cross-platform CRLF/LF issues

### LINKED ISSUE #1952: CRLF issue with StrReplaceFile
### What version of Kimi Code CLI is running?

1.35.0

### Which open platform/subscription were you using?

/login

### Which model were you using?

kimi k2.6

### What platform is your computer?

MacOS

### What issue are you seeing?

When Kimi CLI edits files with CRLF line endings, a small `StrReplaceFile` edit can rewrite the whole file with LF line endings, causing large unrelated diffs.

### What steps can reproduce the bug?

1. Create a repo with Git text normalization disabled:

```bash
mkdir /tmp/kimi-crlf-repro
cd /tmp/kimi-crlf-repro
git init
git config core.autocrlf false
printf '* -text\n' > .gitattributes
printf 'const a = 1;\r\nconst b = 2;\r\n' > test.ts
git add .
git commit -m 'init'
```

2. Run Kimi CLI and ask it to edit one line:

```bash
kimi --work-dir /tmp/kimi-crlf-repro --print --prompt 'Read test.ts and change const b = 2; to const b = 3;'
```

3. Check the diff and bytes:

```bash
git diff -- test.ts
python - <<'PY'
from pathlib import Path
raw = Path("test.ts").read_bytes()
print(repr(raw))
print("CRLF:", raw.count(b"\r\n"))
print("bare LF:", raw.count(b"\n") - raw.count(b"\r\n"))
PY

```

### What is the expected behavior?

Kimi CLI should preserve the original line endings when reading and editing files, so the diff only contains the real content change.
The content change is small, but the file may be rewritten from CRLF to LF, producing noisy line-ending churn.

### Additional information

_No response_

### LINKED ISSUE #2191: [Windows] StrReplaceFile silently converts entire file from CRLF to LF, forcing Agent to abandon native tools for Python workarounds
**Related Issue**: #1952 (CRLF issue with StrReplaceFile)  
**Related PR**: #1953 (fix(file): preserve CRLF during file edits) -- still open/unmerged

---

#### Problem Description

On Windows, repository files use standard `CRLF` (\r\n) line endings (checked out via Git `core.autocrlf=true`). When the Agent attempts to use `StrReplaceFile` for localized edits, it hits a **double failure**:

1. **Using \n (LF) in the `old` string**: The match succeeds, but `StrReplaceFile` **rewrites the entire file from CRLF to LF**, producing a massive unrelated diff.
2. **Using \r\n (CRLF) in the `old` string**: The match fails outright with `"old string was not found"`.

This makes `StrReplaceFile` **completely unsafe to use** on Windows CRLF files. The Agent is forced to fall back to executing Python scripts through the `Shell` tool for every file modification.

---

#### Cascading Consequences (Real Session)

Once forced into Python-via-Shell on Windows PowerShell, a chain of secondary failures occurs:

| Failure | Symptom | Frequency |
|---------|---------|-----------|
| `SyntaxError: EOF while scanning triple-quoted string literal` | Multi-line Python passed via `python -c "..."` gets truncated | Multiple |
| `SyntaxError: EOL while scanning string literal` | Backslash/ newline escaping across PowerShell + Python layers fails | Multiple |
| `SyntaxError: f-string expression part cannot include a backslash` | Windows path backslashes inside f-strings cause syntax errors | Multiple |
| `AttributeError: 'bytes' object has no attribute 'encode'` | Double `.encode()` on a bytes object in the script | Once |
| JSON parse error (`Unterminated string`) | Unescaped quotes inside `SetTodoList` / tool-call JSON payloads | Multiple |
| Tab vs. space mismatch | Some source files mix \t and space indentation; Python scripts searching with the wrong whitespace type fail to find targets | Multiple |

**End result**: Modifying ~7 source files consumed **110 tool calls across 4 conversation turns**, and the files had to be redone from scratch after an accidental revert.

---

#### Reproduction

**Repro 1 -- \n old string silently corrupts line endings**

1. Take any standard Windows CRLF source file (e.g., a C header with 300+ lines, all \r\n).
2. Call `StrReplaceFile` with an `old` string that uses \n for newlines.
3. Replacement succeeds, but the file is now pure LF (\r\n count drops to 0; \n count increases).
4. `git diff` shows the whole file as changed due to line-ending churn.

**Repro 2 -- \r\n old string fails to match**

1. Copy the same CRLF file to a test copy.
2. Call `StrReplaceFile` with an `old` string that explicitly uses \r\n.
3. Returns: `"No replacements were made. The old string was not found in the file."`

---

#### Environment

- **OS**: Windows (case-insensitive filesystem, Git `core.autocrlf=true`)
- **Shell**: PowerShell 5.x / 7.x
- **Kimi CLI Version**: 1.35.0+
- **File format**: UTF-8 (with or without BOM) + CRLF -- standard Windows C source files
- **Project type**: Large embedded project with Git submodules

---

#### Expected Behavior

1. `StrReplaceFile` should **preserve the original line-ending format** of the file; a localized replacement must not rewrite every line ending in the entire file.
2. Alternatively, expose a parameter (e.g., `preserve_line_endings: bool`) to opt-in to preservation.
3. At minimum, when matching fails due to newline mismatch, return **diagnostic detail** (e.g., `"Target text found but newline style mismatch: file uses CRLF, old string uses LF"`) instead of the generic `"old string not found"`.

---

#### Impact Assessment

- **Tool reliability**: Agent can no longer trust `StrReplaceFile` on Windows projects.
- **Efficiency**: A single file edit balloons from 1-2 tool calls to 10+ calls (read -> probe -> script -> verify -> retry).
- **Safety**: Frequent Python syntax errors increase the risk of accidental file corruption; in our session a documentation file was accidentally overwritten by `WriteFile` in overwrite mode because the Agent could not rely on `StrReplaceFile`.
- **UX**: Extremely brittle workflow; user had to interrupt the Agent multiple times and manually restore files.

---

#### Current Workaround

The Agent has effectively abandoned `StrReplaceFile` and now uses:
- **Small files** (<300 lines): `WriteFile` full-file overwrite.
- **Large files** (>=300 lines): Python script via `Shell`, using `open(..., newline='')` to read/write verbatim.

This workaround is highly unstable on PowerShell (quoting hell, multi-layer escaping), so a native fix is urgently needed.

---

#### Suggested Fix Priority

1. **P0**: Internally normalize \n for matching, but write back using the file's original newline sequence.
2. **P1**: Add `preserve_line_endings: bool` parameter to `StrReplaceFile`.
3. **P2**: Return structured diagnostics on mismatch (newline style, indentation character mismatch, BOM presence, etc.).

---

*Full wire.jsonl logs and detailed analysis are retained internally; happy to provide sanitized excerpts if they help with debugging.*

### PR DESCRIPTION
**Pull request by Sisyphbaous-DT-Project**
## Related Issue

Resolve #1952
Resolve #2191

## Description

Fix `StrReplaceFile` and `WriteFile` tools corrupting the original newline style when editing files.

**Root Cause:** `readtext()` uses universal newlines mode by default (`newline=None`), which silently converts `\r\n` to `\n` when reading files; `writetext()` was already fixed in PR #1693 (by adding `newline=""`). This causes the entire read→edit→write pipeline to convert CRLF files to LF, so even if only one line is modified, `git diff` shows the entire file as changed.

**Fix Details:**

1. **`packages/kaos/src/kaos/local.py`**
   - Add `newline=""` to `readtext()` and `readlines()` to disable Python's universal newlines auto-conversion, aligning with the already-fixed `writetext()`.

2. **`packages/kaos/src/kaos/ssh.py`**
   - Change `readtext()` to use `readbytes()` + `decode()` to avoid newline conversion by asyncssh SFTP text mode.
   - Use `splitlines(keepends=True)` in `readlines()` to preserve trailing newlines.
   - Change `writetext()` to byte mode (`"wb"`/`"ab"`) to prevent newline conversion by the remote server.

3. **`src/kimi_cli/tools/file/replace.py`**
   - Change `StrReplaceFile` to read raw bytes and detect the file's dominant newline style (CRLF / LF / CR).
   - Normalize content to `\n` for model matching (model-generated `old`/`new` strings use `\n`).
   - Restore the original newline style after replacement and write to disk via `write_bytes()`.

4. **`src/kimi_cli/tools/file/write.py`**
   - In `WriteFile`, when in overwrite mode and the file already exists, detect its newline style and write the new content in that style.
   - For new files, preserve the newline style of the passed-in content.

**Test Results:**
- `packages/kaos/tests/test_local_kaos.py`: **16 passed**
- Added 9 new EOL-preservation test cases, all passing.
- Additional regression tests: 6/6 passed (covering no-match, replace_all, empty old, append, no-trailing-newline, real-code-scenario).

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/MoonshotAI/kimi-cli/blob/main/CONTRIBUTING.md) document.
- [x] I have linked the related issue, if any.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have run `make gen-changelog` to update the changelog.
- [ ] I have run `make gen-docs` to update the user documentation.

### CODE REVIEW

**Comment by randyxian08** [comment]
There is a line-ending edge case worth covering before merge.

For CRLF files, both the `WriteFile` overwrite path and the `StrReplaceFile` restore path convert with `content.replace("\n", "\r\n")`. If the incoming replacement content already contains CRLF, this turns each `\r\n` into `\r\r\n`.

Example shape:

```py
"a\r\nb\r\n".replace("\n", "\r\n") == "a\r\r\nb\r\r\n"
```

This can happen if the tool input already carries CRLF text, or if `edit.new` contains a CRLF block. A safer approach is to normalize incoming content first (`\r\n`/`\r` -> `\n`) before converting to the detected file EOL, and add a test for pre-CRLF input.

**Comment by Sisyphbaous-DT-Project** [comment]
Thanks to all review bots for your feedback. This commit `5a07501` has fixed the following 3 issues:

**@chatgpt-codex-connector**

- **P1 Badge (write.py: normalize content before converting LF to CRLF)** fixed
  - Added a normalization step before writing: `content.replace("\r\n", "\n").replace("\r", "\n")` to prevent existing CRLF from being double-written into CRCRLF.

- **P2 Badge (replace.py: Preserve mixed line endings during string replacement)** Some improvements
  - Also adds standardized pre-steps to eliminate the risk of double writing.
  - But the global EOL unified transformation is still intentionally simplified by design. Exact line-by-line mapping of mixed line-end files will significantly increase the complexity (previously it was evaluated that the EOL mapping table needs to be established line by line), but mixed line-end files are extremely rare in actual scenarios. The current solution meets most usage scenarios.

- **P2 Badge (ssh.py: returns character count written from SSH text)** fixed
  - Instead `await f.write(encoded)` and then `return len(data)`, making sure to return the number of characters instead of the number of bytes. Added CJK regression test verification.

**@devin-ai-integration**

- **WriteFile append mode does not preserve CRLF line endings** By design
  - The append mode is to append new content. The content passed in by the model is usually the exact text it wants to append. Casting newlines for appended content may not match the model's intent (the model may intentionally append LF lines to a CRLF file). Overwrite requires consistency because it replaces the entire file, while append is an incremental append and is acceptable in edge cases.

- **SSH writetext returns number of bytes** Fixed (ditto)

The above changes have passed all 50 tests (including 12 EOL-related tests).

File: src/kimi_cli/tools/file/write.py [hunk 1]
```diff
@@ -152,10 +163,15 @@ async def __call__(self, params: Params) -> ToolReturnValue:
...
+            # Write content to file preserving original line endings when overwriting
             match params.mode:
                 case "overwrite":
-                    await p.write_text(params.content)
+                    if eol_style != "\n":
+                        # Normalize to LF first to avoid double-converting existing CRLF
+                        content_to_write = params.content.replace("\r\n", "\n").replace("\r", "\n").replace("\n", eol_style)
+                    else:
+                        content_to_write = params.content
+                    await p.write_bytes(content_to_write.encode("utf-8", errors="replace"))
```
[the remark this answers is not in the document]
Comment (Sisyphbaous-DT-Project) [reply src/kimi_cli/tools/file/write.py]: **Status:** Fixed (commit `d2064bb`)

> Removed `errors="replace"` and restored the default `strict` mode. This aligns with the original `write_text()` behavior and is also consistent with the `append` mode. Under `strict` mode, unencodable characters raise an error instead of being silently replaced, adhering to the fail-fast principle.
>
> Note: Since all characters in a Python `str` are valid for UTF-8 encoding, this change produces no observable behavioral difference under normal usage. It primarily eliminates the inconsistency with the original behavior.
